masa/validator/forwarder.py

```diff
@@ -58,33 +61,106 @@ async def forward(
         elif parser_method:
             parsed_responses = parser_method(valid_responses)
 
-        # Score responses
-        rewards = get_rewards(
-            self.validator, query=request.query, responses=parsed_responses
-        )
+        process_times = [synapse.dendrite.process_time for synapse,
+                         uid in zip(synapses, miner_uids) if uid in valid_miner_uids]
 
-        # Update the scores based on the rewards
+        source_of_truth = await self.get_source_of_truth(
+            responses=parsed_responses, miner_uids=miner_uids, source_method=source_method, query=request.query)
 
+        # Score responses
+        rewards = self.get_rewards(responses=parsed_responses, source_of_truth=source_of_truth
+                                   )
+        # Update the scores based on the rewards
         if len(valid_miner_uids) > 0:
             self.validator.update_scores(rewards, valid_miner_uids)
             if self.validator.should_set_weights():
                 try:
                     self.validator.set_weights()
                 except Exception as e:
                     bt.logging.error(f"Failed to set weights: {e}")
-                    
+
         # Add corresponding uid to each response
-        response_with_uids = [
-            {"response": response, "uid": int(uid.item()), "score": score.item()}
-            for response, uid, score in zip(parsed_responses, valid_miner_uids, rewards)
+        responses_with_metadata = [
+            {"response": response, "uid": int(
+                uid.item()), "score": score.item(), "latency": latency}
+            for response, latency, uid, score in zip(parsed_responses, process_times, valid_miner_uids, rewards)
+        ]
+
+        responses_with_metadata.sort(key=lambda x: (-x["score"], x["latency"]))
+        return responses_with_metadata
+
+    def get_rewards(
+        self,
+        responses: dict,
+        source_of_truth: dict
+    ) -> torch.FloatTensor:
+
+        combined_responses = responses.copy()
+        combined_responses.append(source_of_truth)
+
+        embeddings = self.validator.model.encode(
+            [str(response) for response in combined_responses])
+
+        num_clusters = min(len(combined_responses), 2)
```

Comment:
`combined_responses` == `parsed_responses`, no?  this mean that we are grouping / clustering based on the length of responses, meaning each response is in it's own group?

Reply:
parsed_responses is parsed_responses + source_of_truth
not exactly, we are having only 2 groups unless there are less than 2 items, meaning we only group in two groups because grouping in less than that is non sense 😂 
the min thing is only for future settings where we want to rise the number on the right